crates/aptos-substreams/src/transaction_converter.rs

```
@@ -0,0 +1,275 @@
+    }
+}
+
+pub fn get_signature_outputs(
+    s: &Signature,
+    request: &UserTransactionRequest,
+    info: &TransactionInfoOutput,
+) -> Result<Vec<SignatureOutput>> {
+    match s {
+        Signature::Ed255198(sig) => Ok(vec![parse_single_signature(
+            sig, request, info, true, 0, None,
+        )?]),
+        Signature::MultiEd255198(sig) => {
+            Ok(parse_multi_signature(sig, request, info, true, 0, None)?)
+        }
+        Signature::MultiAgent(sig) => Ok(parse_multi_agent_signature(sig, request, info)?),
+    }
+}
+
+fn parse_single_signature(
+    s: &Ed25519Signature,
+    request: &UserTransactionRequest,
+    info: &TransactionInfoOutput,
+    is_sender_primary: bool,
+    multi_agent_index: u32,
+    override_address: Option<&String>,
+) -> Result<SignatureOutput> {
+    let mut signer = request.sender.clone();
+    if let Some(addr) = override_address {
+        signer = addr.clone();
+    }
+    Ok(SignatureOutput {
+        transaction_hash: info.hash.clone(),
+        signer,
+        is_sender_primary,
+        signature_type: get_signature_type(SignatureType::Ed255198),
+        public_key: s.public_key.clone(),
+        signature: s.signature.clone(),
+        threshold: 1,
+        bitmap: vec![],
+        multi_agent_index,
+        multi_sig_index: 0,
+    })
+}
+
+fn parse_multi_signature(
+    s: &MultiEd25519Signature,
+    request: &UserTransactionRequest,
+    info: &TransactionInfoOutput,
+    is_sender_primary: bool,
+    multi_agent_index: u32,
+    override_address: Option<&String>,
+) -> Result<Vec<SignatureOutput>> {
+    let mut signatures = vec![];
+    let mut signer = &request.sender;
+    if let Some(addr) = override_address {
+        signer = addr;
+    }
+    for (index, key) in s.public_keys.iter().enumerate() {
+        let signature = s.signatures.get(index).context(format!(
+            "Failed to get index {} for multi sig signature",
+            index
+        ))?;
```

I think failing hard is potentially okay here: these should always be equal